File: spikes/react-native-typescript-cpp-native/framework-comparison/runtimes/lynx_shell/lynx_shell/android/app/src/main/java/com/lynxshell/modules/OmiNativeModule.kt
package com.lynxshell.modules

import android.content.Context
import android.util.Base64
import com.lynx.jsbridge.LynxMethod
import com.lynx.jsbridge.LynxModule
import org.json.JSONObject

/**
 * Minimal Lynx-to-native seam for the Omi spike.
 * The module loads the shared Omi C++ boundary through a small JNI adapter.
 */
class OmiNativeModule(context: Context) : LynxModule(context) {
    private val ble = OmiBleController(context)
    companion object {
        init {
            System.loadLibrary("omi_lynx_native")
        }
    }

    private external fun nativeCapabilities(): String
    private external fun nativeNormalizePacket(raw: ByteArray): String

    @LynxMethod
    fun getNativeCapabilities(): String = JSONObject(nativeCapabilities())
        .put("framework", "lynx")
        .put("platform", "android")
        .put("bridge", "lynx-native-module")
        .put("cppBoundary", "linked")
        .put("contract", "omi-relay-contract:v1")
        .toString()

    @LynxMethod
    fun normalizePacket(rawBase64: String): String {
        val raw = Base64.decode(rawBase64, Base64.DEFAULT)
        return nativeNormalizePacket(raw)
    }

    @LynxMethod
    fun getBluetoothState(): String = ble.capabilities().toString()

    @LynxMethod
    fun startOmiScan(): String = ble.startScan().toString()

    @LynxMethod
    fun stopOmiScan(): String = ble.stopScan().toString()

    @LynxMethod
    fun getOmiScanResults(): String = ble.scanResults()

    @LynxMethod
    fun connectOmi(address: String): String = ble.connect(address).toString()

    @LynxMethod
    fun disconnectOmi(): String = ble.disconnect().toString()
}

File: spikes/react-native-typescript-cpp-native/framework-comparison/runtimes/lynx_shell/lynx_shell/apple/LynxShell/modules/OmiNativeModule.swift
import Foundation

@objcMembers
public final class OmiNativeModule: NSObject, LynxModule {
  public static var name: String { "OmiNativeModule" }

  public static var methodLookup: [String: String] {
    [
      "getNativeCapabilities": NSStringFromSelector(#selector(getNativeCapabilities)),
      "normalizePacket": NSStringFromSelector(#selector(normalizePacket(_:))),
      "getBluetoothState": NSStringFromSelector(#selector(getBluetoothState)),
      "startOmiScan": NSStringFromSelector(#selector(startOmiScan)),
      "stopOmiScan": NSStringFromSelector(#selector(stopOmiScan)),
      "getOmiScanResults": NSStringFromSelector(#selector(getOmiScanResults)),
      "connectOmi": NSStringFromSelector(#selector(connectOmi(_:))),
      "disconnectOmi": NSStringFromSelector(#selector(disconnectOmi))
    ]
  }

  private let bluetooth = OmiBluetoothController()

  public init(param: Any) { super.init() }
  public override init() { super.init() }

  @objc public func getNativeCapabilities() -> NSString {
    let base = OmiNativeBoundaryBridge.capabilities()
    return "{\"framework\":\"lynx\",\"platform\":\"ios\",\"bridge\":\"lynx-native-module\",\"cppBoundary\":\"linked\",\"capabilities\":\(base)}" as NSString
  }

  @objc public func normalizePacket(_ raw: String) -> NSString {
    OmiNativeBoundaryBridge.normalizePacket(raw) as NSString
  }

  @objc public func getBluetoothState() -> NSString { json(bluetooth.capabilities()) }
  @objc public func startOmiScan() -> NSString { json(bluetooth.startScan()) }
  @objc public func stopOmiScan() -> NSString { json(bluetooth.stopScan()) }
  @objc public func getOmiScanResults() -> NSString { json(bluetooth.scanResults()) }
  @objc public func connectOmi(_ identifier: String) -> NSString { json(bluetooth.connect(identifier: identifier)) }
  @objc public func disconnectOmi() -> NSString { json(bluetooth.disconnect()) }

  private func json(_ value: Any) -> NSString {
    guard JSONSerialization.isValidJSONObject(value),
          let data = try? JSONSerialization.data(withJSONObject: value),
          let text = String(data: data, encoding: .utf8) else { return "{}" }
    return text as NSString
  }
}
